quinn/src/streams.rs

```diff
@@ -482,43 +547,60 @@ where
     fn poll_read_generic<T, U>(
         &mut self,
         cx: &mut Context,
+        ordered: bool,
         mut read_fn: T,
     ) -> Poll<Result<Option<U>, ReadError>>
     where
-        T: FnMut(
-            &mut crate::connection::ConnectionInner<S>,
-            StreamId,
-        ) -> Result<Option<U>, proto::ReadError>,
+        T: FnMut(&mut Chunks) -> (Option<U>, Result<Option<()>, proto::ReadError>),
```

Comment:
This is the resulting change:

```diff
diff --git a/quinn/src/streams.rs b/quinn/src/streams.rs
index 69bd8ffc..c37314bd 100644
--- a/quinn/src/streams.rs
+++ b/quinn/src/streams.rs
@@ -386,7 +386,7 @@ where
                             true => Some(()),
                             false => None,
                         },
-                        Ok(Some(())),
+                        Ok(false),
                     );
                 }
 
@@ -401,7 +401,7 @@ where
                                 true => Some(()),
                                 false => None,
                             },
-                            res.map(|inner| inner.map(|_| ())),
+                            res.map(|inner| inner.is_none()),
                         )
                     }
                 }
@@ -437,8 +437,8 @@ where
         ordered: bool,
     ) -> Poll<Result<Option<Chunk>, ReadError>> {
         self.poll_read_generic(cx, ordered, |chunks| match chunks.next(max_length) {
-            Ok(Some(chunk)) => (Some(chunk), Ok(Some(()))),
-            res => (None, res.map(|inner| inner.map(|_| ()))),
+            Ok(Some(chunk)) => (Some(chunk), Ok(false)),
+            res => (None, res.map(|inner| inner.is_none())),
         })
     }
 
@@ -473,7 +473,7 @@ where
                             0 => None,
                             n => Some(n),
                         },
-                        Ok(Some(())),
+                        Ok(false),
                     );
                 }
 
@@ -488,7 +488,7 @@ where
                                 0 => None,
                                 n => Some(n),
                             },
-                            res.map(|inner| inner.map(|_| ())),
+                            res.map(|inner| inner.is_none()),
                         )
                     }
                 }
@@ -551,7 +559,7 @@ where
         mut read_fn: T,
     ) -> Poll<Result<Option<U>, ReadError>>
     where
-        T: FnMut(&mut Chunks) -> (Option<U>, Result<Option<()>, proto::ReadError>),
+        T: FnMut(&mut Chunks) -> (Option<U>, Result<bool, proto::ReadError>),
     {
         use proto::ReadError::*;
         if self.all_data_read {
@@ -576,8 +584,8 @@ where
         };
 
         match result {
-            Ok(Some(())) => Poll::Ready(Ok(done)),
-            Ok(None) => {
+            Ok(false) => Poll::Ready(Ok(done)),
+            Ok(true) => {
                 self.all_data_read = true;
                 Poll::Ready(Ok(done))
             }
```

I don't think it's a clear win, and I find reasoning about the semantics actually harder because the second element moves away from the established `Option` semantics for stream ending. (Quickly, does `true` mean you can ask the stream for more data or does it mean the stream was done?)

(All the matches are actually about the first element, but that has to actually pack a value to return to its caller.)

Reply:
I guess we could use a struct with named fields for the return value here, but the caller just ends up destructing it immediately (and naming its fields), so it doesn't help that much...

Reply:
One of the things that gave me a lot of pause in the first pass were the nested `map`s, especially as the inner closure will never be invoked since we've separately matched the `Ok(Some(_))` case. We could even replace the inner `map`/`is_none` with a constant, perhaps justified by a comment.

Neither `Option<()>` nor `bool` are particularly readable in and of themselves, even if the former has a correspondence to the ultimate return value. Maybe a special-purpose enum would be a better way to go?

Reply:
Hmm, I like that. What do you think of this approach?